doc/developers/minimal_reproducer.rst

```diff
@@ -0,0 +1,404 @@
+question in the discussions, being able to craft minimal, reproducible examples
+(or minimal, workable examples) is the key to communicating effectively and
+efficiently with the community.
+
+There are very good guidelines on the internet such as `this StackOverflow
+document <https://stackoverflow.com/help/mcve>`_ or `this blogpost by Matthew
+Rocklin <https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports>`_
+on crafting Minimal Complete Verifiable Examples (referred bellow as MCVE).
+Our goal is not to be repetitive with those references but rather to provide a
+step-by-step guide on how to narrow down a bug until you have reached the
+shortest possible code to reproduce it.
+
+The first step before submitting a bug report to scikit-learn is to read the
+`Issue template
+<https://github.com/scikit-learn/scikit-learn/blob/main/.github/ISSUE_TEMPLATE/bug_report.yml>`_.
+It is already quite informative about the information you will be asked to provide.
+
+
+.. _good_practices:
+
+Good practices
+==============
+
+In this section we will focus on the **Steps/Code to Reproduce** section of
+the Issue template. We will start with a snippet of code that already provides a
+failing example but that has room for readability improvement. We then craft a
+MCVE from it.
+
+**Example**
+
+.. code-block:: python
+
+    # I am currently working in a ML project and when I tried to fit a
+    # GradientBoostingRegressor instance to my_data.csv I get a UserWarning:
+    # "X has feature names, but DecisionTreeRegressor was fitted without feature
+    # names". You can get a copy of my dataset from www.this_link.com and verify
+    # my features do have names. The problem seems to arise during fit when I pass
+    # an integer to the n_iter_no_change parameter.
+
+    import pandas as pd
+    df = pd.read_csv('my_data.csv')
+    X = df[["feature_name"]] # my features do have names
+    y = df["target"]
+
+    # We set random_state=42 for the train_test_split
+    from sklearn.model_selection import train_test_split
+    X_train, X_test, y_train, y_test = train_test_split(
+        X, y, test_size=0.33, random_state=42)
+
+    from sklearn.ensemble import GradientBoostingRegressor
+
+    # An instance with default n_iter_no_change raises no error nor warnings
+    gbdt = GradientBoostingRegressor(random_state=0)
+    gbdt.fit(X_train, y_train)
+
+    # the bug appears when I change the value for n_iter_no_change
+    gbdt = GradientBoostingRegressor(random_state=0, n_iter_no_change=5)
+    gbdt.fit(X_train, y_train)
+
+Provide a failing code example with minimal comments
+----------------------------------------------------
+
+Instructions may get lost in translation for non-native english speakers.
+Hence, it is better that all the necessary information can be read
+from the code itself. Besides, by this point you already provided a concise
+description in the **Describe the bug** section of the Issue template.
+
+**Improved example**
+
```

Comment:
```suggestion
The following code, while **still not minimal**, is already **much better** because
it can be copy-pasted in a python terminal to reproduce the problem in one step.
In particular:
- it contains **all necessary imports statements**;
- it can fetch the public dataset without having to manually download a
  file and put it in the expected location on the disk.
```